design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,123 @@
+
+The remainder of this document will propose the addition of lambda functions into ksqlDB’s grammar and execution engine.
+
+## Scope
+
+### What is in scope
+
+* Syntax for defining lambda functions will be proposed
+* Lambda invocation functions will be proposed
+
+## What is not in scope
+
+* Implementation details will not be proposed
+* Level of effort will not be estimated
+
+## Value/return
+
+The introduction of lambda functions will ultimately empower ksqDB users to solve more problems with less effort. Many of the UDFs we’ve seen users write are designed to apply relatively simple functionality to the elements of `Arrays` and `Maps`. Lambda functions are well-suited to solve this class of problems in a very easy and intuitive way.
+
+Lambda functions will also enable some users to use ksqlDB in environments that do not allow for the deployment of UDFs.
+
+## Public APIs
+
+There are two aspects of the user-facing lambda function interface: syntax for describing the actual lambda functions, and a way to specify how to apply these functions to input values.
+
+## Design
+
+### Syntax
+
+It would probably be ideal to use Java-style lambda syntax, but we already use the `->` symbol to represent `Struct` field lookups. Giving `->` a double meaning (if that’s even possible) would create ambiguity and complexity in the grammar/parser. We therefore propose using the `=>` symbol to signify a lambda function:
+
+```
+arg => expr
+```
+
+This expression represents the following:
+
+* `arg` is the argument passed into the scope of each invocation of the lambda function.
+* `expr` is the "body" of the lambda function. Each invocation of the lambda function will evaluate this expression and return the result.
+
+Multiple arguments should also be allowed for lambda functions. When multiple arguments are used, they should be wrapped in parentheses:
+
+```
+(x, y) => expr
+```
+
+#### Allowed expressions
+
+The body of each lambda function should simply be **any expression that could be used in a `SELECT` expression list**, with some exceptions:
+
+* Aggregate function calls (e.g. `sum`, `count`, `avg`) should not be allowed (please see *[Open questions](#open-questions)* section). Note that simple aggregations are still possible with the `reduce_*` invocation functions.
+* `SELECT` subqueries should not be allowed.
+
+#### Examples
+
+Given these restrictions, the following examples would be valid lambda functions:
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
```

Comment:
How about some `flatMap` style transforms?  Not sure what's the best SQLy name, but something where the lambda can return zero or more elements. For want of a better name 'flatmap_array(array, e => e[])`.

This is basically lambdas for table functions.

This could be used, for example, to convert `ARRAY<ARRAY<INT>>` into `ARRAY<INT>` with `flatmap_array(array. e => e)`.

Users can then build more complex UDTFs using CASE for flow and our functions and operators.